src/term/mod.rs

```diff
@@ -131,6 +131,56 @@ pub enum Pattern {
   Tup(Option<Name>, Option<Name>),
 }
 
+impl Pattern {
+  pub fn occurs(&self, name: &Name) -> bool {
+    match self {
+      Pattern::Var(None) => false,
+      Pattern::Var(Some(nam)) => nam == name,
+      Pattern::Ctr(.., args) => {
+        let mut ret = false;
+        for arg in args {
+          ret |= arg.occurs(name);
+        }
+        ret
+      }
+      Pattern::Num(..) => false,
+      Pattern::Tup(fst, snd) => {
+        fst.as_ref().map_or(false, |fst| fst == name) || snd.as_ref().map_or(false, |snd| snd == name)
+      }
+    }
+  }
+
+  pub fn names(&self) -> Vec<&Option<Name>> {
```

Comment:
Why not make a function that returns an iterator over the names? Then you don't need for_each_bind, you can just use `pat.names().for_each()`. To get them into a Vec you can just collect.